server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -0,0 +1,419 @@
+     *
+     * @param timeout The maximum time to wait for the bulk requests to complete
+     * @param unit    The time unit of the {@code timeout} argument
+     * @throws InterruptedException If the current thread is interrupted
+     */
+    public void awaitClose(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
+        synchronized (mutex) {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            if (cancellableFlushTask != null) {
+                cancellableFlushTask.cancel();
+            }
+
+            if (bulkRequest.numberOfActions() > 0) {
+                execute();
+            }
+            this.retry.awaitClose(timeout, unit);
+        }
+    }
+
+    /**
+     * Adds an {@link IndexRequest} to the list of actions to execute. Follows the same behavior of {@link IndexRequest}
+     * (for example, if no id is provided, one will be generated, or usage of the create flag).
+     * @throws EsRejectedExecutionException if adding the approximate size in bytes of the request to totalBytesInFlight would exceed
+     * maxBytesInFlight
+     */
+    public BulkProcessor2 add(IndexRequest request) throws EsRejectedExecutionException {
+        return add((DocWriteRequest<?>) request);
+    }
+
+    /**
+     * Adds an {@link DeleteRequest} to the list of actions to execute.
+     * @throws EsRejectedExecutionException if adding the approximate size in bytes of the request to totalBytesInFlight would exceed
+     * maxBytesInFlight
+     */
+    public BulkProcessor2 add(DeleteRequest request) throws EsRejectedExecutionException {
+        return add((DocWriteRequest<?>) request);
+    }
+
+    /**
+     * Adds either a delete or an index request.
+     * @throws EsRejectedExecutionException if adding the approximate size in bytes of the request to totalBytesInFlight would exceed
+     * maxBytesInFlight
+     */
+    public BulkProcessor2 add(DocWriteRequest<?> request) throws EsRejectedExecutionException {
+        internalAdd(request);
+        return this;
+    }
+
+    /*
+     * Exposed for unit testing
+     */
+    long getTotalBytesInFlight() {
+        return totalBytesInFlight.get();
+    }
+
+    protected void ensureOpen() {
+        if (closed) {
+            throw new IllegalStateException("bulk process already closed");
+        }
+    }
+
+    private void internalAdd(DocWriteRequest<?> request) throws EsRejectedExecutionException {
+        // bulkRequest and instance swapping is not threadsafe, so execute the mutations under a mutex.
+        // once the bulk request is ready to be shipped swap the instance reference unlock and send the local reference to the handler.
+        Tuple<BulkRequest, Long> bulkRequestToExecute = null;
+        synchronized (mutex) {
+            ensureOpen();
+            if (totalBytesInFlight.get() + BulkRequest.getEstimatedSizeInBytes(request) > maxBytesInFlight.getBytes()) {
```

Comment:
Couple more thoughts:

1. Should we flush if this limit trips? It _could_ be that `request` was just unreasonably large, but if we assume the client is fairly well-behaved then we would expect most other calls to fail similarly until we've made some progress on the backlog, and flushing eagerly seems better than waiting for the timeout to elapse.

2. Should we allow one item of grace before tripping the limit? I.e. just check `totalBytesInFlight` vs `maxBytesInFlight`, and always add `request` to the queue if there's at least one byte of space there. I think we can afford that amount of lenience, and it'd mean we can avoid needing to estimate the size of each doc request.

Reply:
For (1) we do flush if this limit trips (we call execute() if newBulkRequestIfNeeded() is not null a few lines below). For (2) I'm changing the code to allow for the lenience you describe, which does simplify things a good bit.

Reply:
Oh sorry, you're right!

Reply:
Alright, I'm changing `newBulkRequestIfNeeded` to return a new bulk request if the bulk request is over its limits (I'm renaming the ambiguous `isOverTheLimit`) or if the totalBytesInFlight >= maxBytesInFlight. So we'll let one request go over a little, then flush it. The next one that comes through will throw an exception (without changing totalBytesInFlight). Since the only way now to increase totalBytesInFlight is to successfully add a request to a bulk, I think this covers it for us (since we do the check/flush after every add).